File: 2022/day-1/lib/day1.ex
defmodule Day1 do
  def part1() do
    File.read!("input")
    |> max_calories()
  end

  def part2() do
    File.read!("input")
    |> sum_top_3()
  end

  def max_calories(lines) do
    get_calories_per_elf(lines)
    |> Enum.max()
  end

  def sum_top_3(lines) do
    get_calories_per_elf(lines)
    |> Enum.sort(:desc)
    |> Enum.take(3)
    |> Enum.sum()
  end

  def get_calories_per_elf(lines) do
    lines
    |> String.split("\n\n", trim: true)
    |> Enum.map(fn elf ->
      String.split(elf, "\n", trim: true)
      |> Enum.map(&String.to_integer(&1))
      |> Enum.sum()
    end)
  end
end

File: 2024/Days/Day1.cs
using System.Collections.Frozen;
using AdventOfCode2024.Lib;

namespace AdventOfCode2024.Days;

public class Day1 : IAdventDay
{
    public string Part1(string input)
    {
        var (left, right) = ParseLists(input);
        
        left.Sort();
        right.Sort();

        var result = left.Zip(right)
            .Select((values) => Math.Abs(values.First - values.Second))
            .Sum();

        return result.ToString();
    }

    public string Part2(string input)
    {
        var (left, right) = ParseLists(input);

        var rightCounts = right.GroupBy(val => val)
            .Select(group => KeyValuePair.Create(group.Key, group.Count()))
            .ToFrozenDictionary();
        
        var result = left.Aggregate(0, (current, val) => current + rightCounts.GetValueOrDefault(val, 0) * val);

        return result.ToString();
    }
    
    private (List<int>, List<int>) ParseLists(string input)
    {
        var inputSpan = input.AsSpan();
        var lines = inputSpan.Split("\n");
        var left = new List<int>(1000);
        var right = new List<int>(1000);
        
        foreach (var line in lines)
        {
            left.Add(int.Parse(inputSpan[line][..5]));
            right.Add(int.Parse(inputSpan[line][8..]));
        }
        
        return (left, right);
    }
}